File: golang-devildyw/middle/L39组合总和.go
package main

func main() {}

func combinationSum(candidates []int, target int) [][]int {
	result := make([][]int, 0)
	path := make([]int, 0, len(candidates)) // 预分配一点容量
	var backtrack func(index int, sum int)
	backtrack = func(index int, sum int) {
		if sum == target {
			temp := make([]int, len(path))
			copy(temp, path)
			result = append(result, temp)
			return
		}
		if sum > target {
			return
		}

		for i := index; i < len(candidates); i++ {
			path = append(path, candidates[i])
			backtrack(i, sum+candidates[i])
			path = path[:len(path)-1]
		}
	}
	backtrack(0, 0)
	return result
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L39组合总和.java
package top.dyw.leetcode.Hot100;

public class L39组合总和 {
    public List<List<Integer>> result = new ArrayList<>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        dfs(candidates, target, new ArrayList<>(), 0, 0);
        return result;
    }

    public void dfs(int[] candidates, int target, List<Integer> path, int curSum, int index) {
        if (curSum == target) {
            result.add(new ArrayList(path));
            return;
        }
        if (curSum>target) {
            return;
        }

        for (int i=index; i<candidates.length; i++) {
            path.add(candidates[i]);
            dfs(candidates, target, path, curSum+candidates[i], i);
            path.remove(path.size()-1);
        }
    }
}
